File: implementation/swift/GameEventScriptTool/Sources/GameEventScriptTool/RunSession.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

import Foundation
import GameEventScriptRuntime

final class RunSession {
    private(set) var host: GameEventScriptHost
    private(set) var observer: RunObserver
    private(set) var inventory: RunInventory
    let io: ToolIO
    let options: RunOptions
    var messages = 0, opcodes = 0, emits = 0, publishes = 0

    init(options: RunOptions, io: ToolIO, nextID: Int = 1) throws {
        self.options = options
        self.io = io
        observer = RunObserver(io: io, options: options)
        host = try GameEventScriptHost(seed: options.seed, limits: options.limits, observer: observer)
        inventory = RunInventory(host: host, io: io, nextID: nextID)
        try observer.subscribe(inventory)
    }

    func reload(activePath: inout String) throws -> Bool {
        let replacement = try RunSession(options: options, io: io, nextID: inventory.nextID)
        // Link the complete group before initialization; preparation failures leave this session intact.
        for entry in inventory.active {
            activePath = entry.paths[0]
            let program = try entry.paths.count == 1 ? ToolFiles.read(entry.paths[0]) : ToolFiles.compile(entry.paths)
            try replacement.inventory.load(program, paths: entry.paths, id: entry.id)
        }
        guard try replacement.pump() else { return false }
        inventory.unloadAll()
        host = replacement.host
        observer = replacement.observer
        inventory = replacement.inventory
        messages += replacement.messages
        opcodes += replacement.opcodes
        emits += replacement.emits
        publishes += replacement.publishes
        return true
    }

    func pump() throws -> Bool {
        if !host.isReady {
            let start = try host.start()
            messages += start.processedMessages
            opcodes += start.executedOpcodes
            emits += start.emittedMessages
            publishes += start.publishedMessages
            guard start.state == .ready else { return false }
        }
        let result = try host.runToCompletion()
        messages += result.processedMessages
        opcodes += result.executedOpcodes
        emits += result.emittedMessages
        publishes += result.publishedMessages
        return !observer.failed && io.outputError == nil && result.state == .completed
    }

    func summary() { io.line("Run completed: \(messages) messages, \(opcodes) opcodes, \(emits) emits, \(publishes) publishes.", toError: true) }
}

final class RunObserver: GameEventScriptRuntimeObserver {
    let io: ToolIO
    let options: RunOptions
    private lazy var highlighting = Highlighting()
    var failed = false
    var exitCode = 0

    init(io: ToolIO, options: RunOptions) {
        self.io = io
        self.options = options
    }

    func subscribe(_ inventory: RunInventory) throws {
        try inventory.subscribe("ConsoleOut", handler: ConsoleHandler(observer: self, error: false))
        try inventory.subscribe("ConsoleErr", handler: ConsoleHandler(observer: self, error: true))
        try inventory.subscribe("ErrorCode", handler: ErrorCodeHandler(observer: self))
    }

    func messageEmitted(_ message: GameEventScriptMessage, accepted: Bool) { trace("emit " + format(message) + (accepted ? "" : " [not queued]")) }

    func messagePublished(_ message: GameEventScriptMessage, result: GameEventScriptPublishResult) { trace("publish " + format(message) + (result.localAccepted ? "" : " [not queued]")) }

    func dispatchStarted(_ message: GameEventScriptMessage, signatureID: String) { trace("dispatch \(signatureID) <- " + format(message)) }

    func runtimeLimitReached(_ limitName: String, detail: String, limit: Int) {
        failed = true
        io.line("error cli.runtimeLimit: \(limitName)=\(limit): \(detail)", toError: true)
    }

    func runtimeError(_ diagnostic: GameEventScriptDiagnostic) {
        failed = true
        io.diagnostic(diagnostic, fallback: "runtime")
    }

    private func trace(_ text: String) { if options.verbose { io.line(options.color && options.interactive ? Highlighting.paint(text, 33) : text, toError: true) } }

    private func format(_ message: GameEventScriptMessage) -> String {
        var text = message.name
        if !message.arguments.isEmpty {
            text += "(" + (0..<message.arguments.count).map { message.arguments.nameAt($0) + ": " + (message.arguments[$0].kind == .text ? quoted(message.arguments[$0].asText) : message.arguments[$0].description) }.joined(separator: ", ") + ")"
        }
        if !message.tags.isEmpty { text += " with " + message.tags.map { "#" + $0 }.joined(separator: ", ") }
        return text
    }

    func console(_ message: GameEventScriptMessage, error: Bool) {
        var text = ""
        for index in 0..<message.arguments.count {
            let value = message.arguments[index]
            if options.color && !error && value.kind != .text { text += [.integer, .float, .percentage].contains(value.kind) ? Highlighting.paint(value.description, 34) : highlighting.render(value.description) } else { text += value.description }
        }
        io.line(options.color && error ? Highlighting.paint(text, 31) : text, toError: error)
    }

    private struct ConsoleHandler: GameEventScriptNativeMessageHandler {
        let observer: RunObserver
        let error: Bool

        func handle(_ message: GameEventScriptMessage, context: GameEventScriptContext) throws { observer.console(message, error: error) }
    }

    private struct ErrorCodeHandler: GameEventScriptNativeMessageHandler {
        let observer: RunObserver

        func handle(_ message: GameEventScriptMessage, context: GameEventScriptContext) throws {
            guard message.arguments.count == 1, ["code", "_"].contains(message.arguments.nameAt(0)) else { throw try invalid() }
            let value = message.arguments[0]
            if value.kind == .nothing {
                observer.exitCode = 0
                return
            }
            let number = value.asNumber
            guard [.integer, .float].contains(value.kind), value.unit == .none, number.isFinite, (0...255).contains(number), number == number.rounded(.towardZero) else { throw try invalid() }
            observer.exitCode = Int(number)
        }

        private func invalid() throws -> GameEventScriptExtensionFault { try .init(code: "cli.errorCodeArgument", message: "ErrorCode requires one code argument: a unitless integer from 0 to 255, or nothing.") }
    }
}

File: implementation/csharp/GameEventScript.Tool/src/RunSession.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using GameEventScript.Api;

namespace GameEventScript.Tool;

internal sealed class RunSession
{
    private readonly long? _seed;
    private readonly GameEventScriptRuntimeLimits _limits;
    private readonly bool _verbose, _color, _interactive;
    internal GameEventScriptHost Host { get; private set; }
    internal RunObserver Observer { get; private set; }
    internal RunInventory Inventory { get; private set; }

    internal RunSession(long? seed, GameEventScriptRuntimeLimits limits, bool verbose, bool color, bool interactive, int nextId = 1)
    {
        _seed = seed;
        _limits = limits;
        _verbose = verbose;
        _color = color;
        _interactive = interactive;
        Observer = new RunObserver(verbose, color, interactive);
        var builder = GameEventScriptHost.CreateBuilder().WithRuntimeObserver(Observer).WithRuntimeLimits(limits);
        if (seed is { } configuredSeed) builder.WithRandomSeed(configuredSeed);
        Host = builder.Build();
        Inventory = new RunInventory(Host, nextId);
        Observer.SubscribeConsoleHandlers(Inventory);
    }

    internal bool Reload(ref string? activePath)
    {
        var replacement = new RunSession(_seed, _limits, _verbose, _color, _interactive, Inventory.NextId);
        // Link the entire group before initialization, preserving the old session on preparation failure.
        foreach (var entry in Inventory.ActivePrograms())
        {
            activePath = entry.Paths[0];
            var program = entry.Paths.Length == 1
                ? RunProgramFiles.ReadOne(activePath)
                : RunProgramFiles.Compile(entry.Paths, ref activePath);
            replacement.Inventory.Load(program, entry.Paths, entry.Id);
        }
        if (!replacement.Pump()) return false;
        Inventory.UnloadAll();
        Host = replacement.Host;
        Observer = replacement.Observer;
        Inventory = replacement.Inventory;
        _processedMessages += replacement._processedMessages;
        _opcodes += replacement._opcodes;
        _emits += replacement._emits;
        _publishes += replacement._publishes;
        return true;
    }

    private long _processedMessages;
    private long _opcodes;
    private long _emits;
    private long _publishes;

    internal bool Pump()
    {
        if (!Host.IsReady)
        {
            var start = Host.Start();
            _processedMessages += start.ProcessedMessages;
            _opcodes += start.ExecutedOpcodes;
            _emits += start.EmittedMessages;
            _publishes += start.PublishedMessages;
            if (start.State != GameEventScriptStartState.Ready) return false;
        }
        GameEventScriptExecutionResult result;
        do
        {
            result = Host.RunToCompletion();
            _processedMessages += result.ProcessedMessages;
            _opcodes += result.ExecutedOpcodes;
            _emits += result.EmittedMessages;
            _publishes += result.PublishedMessages;
            if (Observer.OutputError is { } outputError)
            {
                Console.Error.WriteLine($"error cli.io: {outputError.Message}");
                return false;
            }
            if (Observer.Failed) return false;
            if (!_interactive && result.State == GameEventScriptExecutionState.Waiting && Host.NextMessageDelay is { } delay)
                Thread.Sleep((int)Math.Min(1000, delay / 1000 + (delay % 1000 == 0 ? 0 : 1)));
        } while (!_interactive && result.State == GameEventScriptExecutionState.Waiting);
        return result.State is GameEventScriptExecutionState.Completed or GameEventScriptExecutionState.Waiting;
    }

    internal void WriteSummary()
        => Console.Error.WriteLine($"Run completed: {_processedMessages} messages, {_opcodes} opcodes, {_emits} emits, {_publishes} publishes.");
}
